lite/examples/sound_classification/ios/SoundClassification/GMLAudioRecord/GMLAudioRecord.h

```diff
@@ -0,0 +1,67 @@
+// Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import "GMLAudioFormat.h"
+#import "GMLRingBuffer.h"
+
+@import AVFoundation;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/** A wrapper class to tap the device's microphone continuously. Currently this class only supports
+ tapping the input node of the AVAudioEngine which emits audio data having only one channel.*/
+NS_SWIFT_NAME(AudioRecord)
+@interface GMLAudioRecord : NSObject
+
+/** Audio format specifying the number of channels and sample rate supported. */
+@property(nonatomic, readonly) GMLAudioFormat *audioFormat;
+
+/** Size of the buffer held by GMLAudioRecord. It ensures delivery of audio data of length bufferSize
+ * arrays when you tap the input microphone. */
+@property(nonatomic, readonly) NSUInteger bufferSize;
+
+/**
+ * Initializes GMLAudioRecord with a GMLAudioFormat and sample count.
+ *
+ * @param format An audio format of type GMLAudioFormat.
+ * @seealso GMLAudioFormat
+ *
+ * @param sampleCount The number of samples this GMLAudioRecord instance should delliver
+ * continuously when you tap the on-device microphone. The tap callback will deliver arrays of size
+ * sampleCount * bufferSize when you tap the microphone using
+ * (checkAndStartTappingMicrophoneWithCompletionHandler:).
+ *
+ * @return An instance of GMLAudioRecord
+ */
+- (nullable instancetype)initWithAudioFormat:(GMLAudioFormat *)format
+                                 sampleCount:(NSUInteger)sampleCount
+                                       error:(NSError *_Nullable *)error;
+
+/**
+ * Taps the input of the on-device microphone and delivers the incoming audio data continuously in a
+ * completion handler. Note that the completion handler delivers results on a background thread.
+ *
+ * @param completionHandler Completion handler deliivers either a buffer of size bufferSize or an
+ * error failing to do so .
+ *
+ */
+- (void)checkPermissionsAndStartTappingMicrophoneWithCompletionHandler:
```

Comment:
Please rename to `startRecordingWithError` and `stop` to align with the Python and Android API.

The TFLAudioRecord should have a getter `buffer` to return the internal buffer, rather than returning the internal buffer only when start recording.

Reply:
The getter will have a completion handler then. Because iOS AudioEngine returns data in callback continuously. Audio callbacks in iOS is always returned in a background thread. So we will have to ensure reads and writes happen in the same thread as reading from a different thread will return out of sync data. For this a dispatch_barrier_async will have to be used which won't return immediately.
So if we cut out the completion handler for data delivery here, we will have a completion handler in read. Also we won't be able to keep track of errors in real time. If there is any error we will have to populate the buffer with -1s or something and leave it to the users to figure out the issue. The read method can't return a status too as read will happen in a callback.

Reply:
You can use a GCD's serial queue to make sure that read and write to the AudioRecord's internal buffer don't happen at the same time. In the getter, you can use GCD's sync method to wait for reading from the internal's queue to complete before returning the copy of the buffer values in the correct order to the caller.

Reply:
I have fixed this according to your suggestion. Do have a look